[scan] Autograd

bohnstingl opened the pull request:
This is the new PR to realize autograd for scan. This is a derivative of https://github.com/bohnstingl/pytorch/pull/2

@ydwu4 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

bohnstingl committed:
First working version of Autograd for scan

bohnstingl committed:
Merge branch 'gh/ydwu4/153/head' of github.com:pytorch/pytorch into scan_autograd2

bohnstingl committed:
WIP: Autograd currently fails for some compile modes and the reverse flag

Skylion007 commented on torch/_inductor/ir.py:
```diff
@@ -6745,7 +6745,7 @@ def extract_scan_args(
             *V.graph.current_node.args
         )
-        fx_all_inputs = fx_init + fx_xs + fx_additional_inputs
+        fx_all_inputs = fx_init + fx_xs + tuple(fx_additional_inputs) if type(fx_init) == tuple else fx_additional_inputs
```
```suggestion
        fx_all_inputs = fx_init + fx_xs + tuple(fx_additional_inputs) if type(fx_init) is tuple else fx_additional_inputs
```

bohnstingl replied:
This has been resolved in a different way

bohnstingl committed:
Minor fixes

bohnstingl committed:
Merge branch 'gh/ydwu4/153/head' of github.com:pytorch/pytorch into scan_autograd2

bohnstingl committed:
WIP: Integration of scan lowering fixes

bohnstingl committed:
WIP: Autograd test updates

bohnstingl committed:
Minor cleanup

ydwu4 reviewed (commented):
Overall we're on the right track! Add a few comments. The additonal_inputs would need some special handling and some more testing.

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -360,9 +410,153 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
+def scan_autograd(combine_fn, init, input, dim, reverse, additional_inputs):
+    # A shortcut for the case where all inputs don't require gradient,
+    # we skip tracing the forward and backward graph.
+    if pytree.tree_all_only(
+        torch.Tensor,
+        lambda t: not t.requires_grad,  # type: ignore[union-attr]
+        (init, input),
+    ):
+        with torch._C._AutoDispatchBelowAutograd():
+            return scan_op(combine_fn, init, input, dim, reverse, additional_inputs)
```
Currently it does, I can see this in test. But if this the DispatchKey is not invoked at all, this would also be an option

bohnstingl replied:
I added a TODO for this

bohnstingl committed:
Merge branch 'gh/ydwu4/153/head' of github.com:pytorch/pytorch into scan_autograd2

bohnstingl committed:
Updated implementation of scan autograd and integrated comments from code review

bohnstingl committed:
Updated scan autograd

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -42,9 +58,62 @@ def wrap_combine_fn_flat(
     assert num_init_leaves == len(carry_flat)
     return [*carry_flat, *combined_flat]
 
+def create_fw_bw_graph_combinefn(combine_fn, init, input, dim):
```
Done

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -360,9 +410,153 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
+        init = operands_outs[:num_leaves_init]
+        xs = operands_outs[num_leaves_init : num_leaves_init + num_leaves_xs]
+        carries = operands_outs[
+            num_leaves_init + num_leaves_xs : 2 * num_leaves_init + num_leaves_xs
```
Done

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -360,9 +410,153 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
+                ]
+                xs = [torch.flip(x, [dim]) for x in xs]
+            else:
+                carries_augmented = [
```
Done

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -360,9 +410,153 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
+            )
+
+            if reverse:
+                g_xs = [torch.flip(g, [dim]) for g in g_xs]
```
Done

bohnstingl committed:
Fixed issues some issues with autograd on non-zero scan dim

bohnstingl committed:
Added functionality to provide gradients for additional inputs (closures)

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -360,9 +410,153 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
+    (
+        fw_graph,
+        joint_graph,
+    ) = create_fw_bw_graph_combinefn(combine_fn, init, input, dim)
```
I implemented a version of this and also modified the RNN testcase to demonstrate that

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
+    if pytree.tree_all_only(
+        torch.Tensor,
+        lambda t: not t.requires_grad,  # type: ignore[union-attr]
+        (init, xs),
```
We also need to check additional_inputs?

bohnstingl replied:
Done

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
+    num_leaves_init = len(init)
+    num_leaves_additional_inputs = len(additional_inputs)
+
+    if not torch.is_grad_enabled():
```
This worths a comment on when this could happen: e.g. in the double-backward case when we trigger the autogradKey of backward. Can merge with the previous TODO.

probably use a different variable name like new_dim and add a comment on why we force it to be 0.

bohnstingl replied:
Done

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
+        num_leaves_additional_inputs,
+        *ops,
+    ):
+        ops_list = list(ops)
+        init = ops_list[:num_leaves_init]
+        xs = ops_list[num_leaves_init : len(ops_list) - num_leaves_additional_inputs]
+        additional_inputs = ops_list[len(ops_list) - num_leaves_additional_inputs :]
```
Maybe a locally scoped helper function:
`def split_operands(ops, num_leaves_init, num_leaves_additional_inputs)`

nit: ops usually refers to operators, so probably change it flat_args or something

bohnstingl replied:
Done, I simplified and unified the code

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -4,32 +4,93 @@
+    return t.permute(*dims)
+
+
+def prepare_xs_carries_for_bwd(xs, init, carries, dim, reverse):
```
We don't expect two prepare_* functions to be used by other file, right? So maybe move these two prepare function inside of backward to make it locally scoped? They are not exposed to users in this way.

bohnstingl replied:
I moved the functions inside the backward

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -43,8 +104,64 @@ def wrap_combine_fn_flat(
+def create_fw_bw_graph_combinefn(combine_fn, init, xs, dim, additional_inputs):
+    # See Note [HOP create fw_bw graph] in create_fw_bw_graph in utils.py
+
+    # Helper wrapper for the autograd forward.
```
Can add a comment 
```
# TODO: do less re-computation with min-cut partitioner. 
```

bohnstingl replied:
Done

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -43,8 +104,64 @@ def wrap_combine_fn_flat(
+                    f"Got types {[type(out) for out in fw_outputs]}."
+                )
+
+            # TODO: There is a major issue that the create_fw_bw in the higher_order_op is invoked twice:
```
We can probably move to where we do the torch.is_grad_enabled() check (need to change the wording a bit).

bohnstingl replied:
Done

ydwu4 commented on torch/_higher_order_ops/utils.py:
```diff
@@ -264,10 +264,12 @@ def maybe_clone(t):
     def fw_with_masks(*args):
         fw_out = fn(*args)
-        return fw_out, [
+        fw_out_flat, tree_spec = pytree.tree_flatten(fw_out)
```
Do we need this change?

bohnstingl replied:
Good catch, no we don't. This was an artifact from resolving earlier issues

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
+        It does so by constructing using an additional scan operator with the gradients
+
+        Args:
+            flat_grads (torch.Tensor): The tensor of upstream gradients, or a nested pytree of tensors.
```
I believe they're always flattened  with no pytree involved.

bohnstingl replied:
Corrected

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
+        Example::
+
+            The ``fw_graph`` f(.,.), used in the forward function, is the operator used during the scan. For example
+            def f(x: torch.Tensor, y: torch.Tensor):
```
this example is a bit boring: maybe change to next_carry = y = x * y?

bohnstingl replied:
I rewrote the example for the multiply function

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -110,9 +227,6 @@ def add(x: torch.Tensor, y: torch.Tensor):
     if not isinstance(reverse, bool):
         raise RuntimeError("Reverse must be a bool, but got " + str(type(reverse)))
 
-    # TODO: Support closures/nn_modules in order to be able represent RNNs with scan
```
Don't need to happen in this PR: it seems to me that can we handle the dim kwarg in scan (i.e. the user-facing wrapper) by shift it backward to dim 0 so that we could remove all the dim arguments in scan_op. This would simplify everything related to dim. WDYT? Maybe add a TODO?

bohnstingl replied:
I am not so sure, it may be that the `combine_fn` depends internally on the shape that the user provides, right?
One concrete example:
```
W = torch.randn(3, 5)
xs = torch.randn(5, 7, 3)
init = torch.randn(3, 3)
def f(x, y):
  return W @ x + y, W @ x + y
carry, out = scan(f, init, xs, 1)
```
In this case we cannot always move the dim to 0, as the `matmul` would break, right?

ydwu4 replied:
like we could do scan(f, init, xs_shifted, 0), where xs_shifted.shape = (7, 5, 3)

bohnstingl replied:
True, I'll add a TODO for that

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -2459,8 +2527,16 @@ def add_scalar_carry(x: torch.Tensor, y: torch.Tensor):
         self.assertEqual(result_init, result_exp)
         self.assertEqual(result_init[0], torch.tensor([3.0], device=device))
 
+        if autograd:
```
all the autograd tests looks similar, can we extract them out as a funciton: check_autograd?

bohnstingl replied:
I factored out most of them

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
+            g_xs = g_xs[:num_leaves_xs]
+            g_xs = prepare_final_gradients_xs(g_xs, dim, reverse)
+
+        return *[None] * 6, *g_init, *g_xs, *new_g_additional_inputs
```
Do we have tests for testing gradient of additional_inputs? I feel we need to modify the backward graph as well:
```
def new_joint_graph(carried_grad_additional_inputs, ...):
   ...grad_additioal_inptus... = joint_graph(...)
   carry (..., grad_addtioanl_inputs + carried_grad_additional_inputs),
```
The init should be a torch.zeros() note that: the joint_graph doesn't take carried_grad_additional_inputs as carry. it only takes additioanl_inputs as part of xs. It's the wrapper new_joint_graph that takes a new carry.

bohnstingl replied:
I've included a wrapper code that takes care of the gradients for the additional inputs

bohnstingl committed:
Code cleanup and integration of review comments

bohnstingl committed:
WIP: autograd for nested additional_inputs

bohnstingl committed:
WIP: Minor updates

bohnstingl committed:
Fixed lintrunner issues

bohnstingl committed:
Working version of autograd

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
+            flat_grads_list = list(flat_grads)
+            g_c_T = flat_grads_list[:num_leaves_init]
+            g_ys = flat_grads_list[num_leaves_init : num_leaves_init + num_leaves_ys]
+            old_g_additional_inputs_T = flat_grads_list[
```
I've added an additional test to ensure that the gradients for the additional inputs are correct

bohnstingl committed:
Fixed bugs and testcase fails

bohnstingl committed:
Merge branch 'gh/ydwu4/153/head' of github.com:pytorch/pytorch into scan_autograd2

bohnstingl committed:
Added a few TODOs
Added checks for the case if not all variables in Autograd require gradients. Not yet handled

bohnstingl commented:
@pytorchbot add label "no-stale"

bohnstingl commented:
@pytorchbot label "no-stale"

bohnstingl commented:
This PR will be combined with https://github.com/bohnstingl/pytorch/pull/4 and pursued further in https://github.com/pytorch/pytorch/pull/146285
